File: examples/typescript/09_streaming_consumer.ts
// (C) Copyright 2026- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
// In applying this licence, ECMWF does not waive the privileges and immunities
// granted to it by virtue of its status as an intergovernmental organisation nor
// does it submit to any jurisdiction.

/**
 * Example 09 — Streaming consumer (TypeScript)
 *
 * Decode tensogram messages as bytes arrive on a `ReadableStream`
 * without ever buffering the full file.  On every chunk the consumer
 * calls `scan(buf)` on its rolling buffer, decodes each completed
 * message, and drops the consumed prefix — peak memory is one complete
 * message, not the whole stream.
 *
 * This is the explicit-loop counterpart to {@link decodeStream} (see
 * `05_streaming_fetch.ts`).  Use `decodeStream` when you want the
 * library to manage the buffer internally; use this pattern when the
 * caller needs direct visibility into message boundaries (progress
 * bars, per-message cancellation, mixing with other protocols on the
 * same byte stream, ...).
 *
 * The example synthesises a stream in-process so it runs anywhere;
 * point `streamOf(...)` at `(await fetch(url)).body!` to drive the
 * same loop off a real HTTP response.
 */

import {
  decode,
  encode,
  init,
  scan,
  type DataObjectDescriptor,
  type GlobalMetadata,
} from '@ecmwf.int/tensogram';

function describe(shape: number[]): DataObjectDescriptor {
  const strides = new Array<number>(shape.length).fill(1);
  for (let i = shape.length - 2; i >= 0; i--) strides[i] = strides[i + 1] * shape[i + 1];
  return {
    type: 'ntensor',
    ndim: shape.length,
    shape,
    strides,
    dtype: 'float32',
    byte_order: 'little',
    encoding: 'none',
    filter: 'none',
    compression: 'none',
  };
}

function buildFourMessages(): Uint8Array {
  const params: Array<[string, number]> = [
    ['2t', 0],
    ['10u', 0],
    ['10v', 0],
    ['msl', 0],
  ];
  const parts: Uint8Array[] = [];
  for (const [param, step] of params) {
    const metadata: GlobalMetadata = {
      version: 3,
      base: [{ mars: { param, step } }],
    };
    // 181×360 float32 — a realistic gridded field.
    const data = new Float32Array(181 * 360).fill(Math.random());
    parts.push(encode(metadata, [{ descriptor: describe([181, 360]), data }]));
  }
  const total = parts.reduce((n, p) => n + p.byteLength, 0);
  const out = new Uint8Array(total);
  let o = 0;
  for (const p of parts) {
    out.set(p, o);
    o += p.byteLength;
  }
  return out;
}

/** Feed bytes into a ReadableStream split into small chunks. */
function streamOf(bytes: Uint8Array, chunkSize = 4096): ReadableStream<Uint8Array> {
  let offset = 0;
  return new ReadableStream({
    pull(controller) {
      if (offset >= bytes.byteLength) {
        controller.close();
        return;
      }
      const end = Math.min(offset + chunkSize, bytes.byteLength);
      controller.enqueue(bytes.subarray(offset, end));
      offset = end;
    },
  });
}

async function main(): Promise<void> {
  await init();

  const fileBytes = buildFourMessages();
  console.log(`Built ${fileBytes.byteLength} bytes over 4 messages\n`);

  const stream = streamOf(fileBytes, 4096);
  const reader = stream.getReader();

  // Rolling buffer: grows with each chunk, shrinks as complete
  // messages are consumed.  Its peak length is at most one message.
  let buffer = new Uint8Array(0);
  let messagesDecoded = 0;

  console.log('Streaming (chunk size = 4096):');
  console.log('─'.repeat(60));

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    // Extend the rolling buffer with the new chunk.
    const merged = new Uint8Array(buffer.byteLength + value.byteLength);
    merged.set(buffer);
    merged.set(value, buffer.byteLength);
    buffer = merged;

    // scan tolerates a partial trailing message — it only returns
    // positions it could verify start-to-end.
    const positions = scan(buffer);
    let consumed = 0;
    for (const { offset, length } of positions) {
      const slice = buffer.subarray(offset, offset + length);
      const msg = decode(slice);
      try {
        const param = (msg.metadata.base?.[0] as { mars?: { param: string } } | undefined)
          ?.mars?.param ?? '?';
        messagesDecoded += 1;
        console.log(
          `  message ${messagesDecoded}: param=${param.padEnd(3)}  ` +
          `shape=[${msg.objects[0].descriptor.shape.join(', ')}]  ` +
          `${slice.byteLength} bytes`,
        );
      } finally {
        msg.close();
      }
      consumed = offset + length;
    }

    // Drop the consumed prefix; keep any partial-message tail.
    if (consumed > 0) {
      buffer = buffer.subarray(consumed);
    }
  }

  console.log('─'.repeat(60));
  console.log(`Total: ${messagesDecoded} messages decoded from stream`);
  console.log('Peak buffer: single message (not whole file)');

  if (messagesDecoded !== 4) {
    throw new Error(`expected 4 messages, decoded ${messagesDecoded}`);
  }
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});

File: examples/cpp/09_streaming_consumer.cpp
// (C) Copyright 2026- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
// In applying this licence, ECMWF does not waive the privileges and immunities
// granted to it by virtue of its status as an intergovernmental organisation nor
// does it submit to any jurisdiction.

/// @file 09_streaming_consumer.cpp
/// @brief Example 09 — Streaming consumer: decode tensograms as bytes arrive.
///
/// Demonstrates consumer-side streaming: read a `.tgm` byte stream in
/// small chunks, scan the growing buffer for complete messages on each
/// chunk boundary, decode each completed message immediately, and
/// discard consumed bytes.  Peak buffer is a single message, not the
/// whole file.
///
/// A real-world caller would read the chunks from an HTTP response, a
/// socket, or a pipe.  This example drives the same pattern off a
/// local `.tgm` file to keep the example self-contained.
///
/// Build:
///   cmake -S cpp -B build -DCMAKE_BUILD_TYPE=Release
///   cmake --build build -j
///   ./build/bin/09_streaming_consumer

#include <tensogram.hpp>

#include <array>
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <filesystem>
#include <fstream>
#include <string>
#include <vector>

#include <unistd.h>

static std::string descriptor_json(const char* param) {
    char buf[512];
    std::snprintf(buf, sizeof(buf),
        R"({"version":3,"descriptors":[{"type":"ntensor","ndim":2,"shape":[181,360],"strides":[360,1],"dtype":"float32","encoding":"none","filter":"none","compression":"none"}],"base":[{"mars":{"param":"%s","step":0}}]})",
        param);
    return buf;
}

int main() {
    const auto tmp = std::filesystem::temp_directory_path() /
                     ("tensogram_example_09_" + std::to_string(::getpid()) + ".tgm");
    const std::string tgm_path_str = tmp.string();
    const char* tgm_path = tgm_path_str.c_str();
    const char* PARAMS[] = {"2t", "10u", "10v", "msl"};

    // ── 1. Create a multi-message .tgm file on disk ───────────────────────
    {
        auto f = tensogram::file::create(tgm_path);
        std::vector<float> data(181 * 360, 0.0f);  // filled below per param
        for (std::size_t i = 0; i < std::size(PARAMS); ++i) {
            // Distinct sentinel value per param so the consumer can verify
            // the right payload landed against the right metadata.
            for (auto& v : data) v = static_cast<float>(i) + 0.5f;
            std::vector<std::pair<const std::uint8_t*, std::size_t>> objects = {
                {reinterpret_cast<const std::uint8_t*>(data.data()),
                 data.size() * sizeof(float)}
            };
            f.append(descriptor_json(PARAMS[i]), objects);
        }
    }
    std::ifstream probe(tgm_path, std::ios::binary | std::ios::ate);
    const auto total_bytes = static_cast<std::size_t>(probe.tellg());
    probe.close();
    std::printf("Wrote %s: %zu bytes, %zu messages\n",
                tgm_path, total_bytes, std::size(PARAMS));

    // ── 2. Stream from disk in small chunks, decode as messages complete ──
    //
    // Strategy: grow a rolling buffer, call scan() on each new chunk,
    // decode any messages scan reports, then drop consumed bytes from
    // the head of the buffer.  scan() tolerates a partial message at
    // the buffer tail — it just won't appear in the result until the
    // last byte arrives.
    std::printf("\nStreaming from %s (chunk size = 4096)\n", tgm_path);
    std::printf("%s\n", std::string(60, '-').c_str());

    std::ifstream in(tgm_path, std::ios::binary);
    std::vector<std::uint8_t> buffer;
    std::array<char, 4096> chunk{};
    std::size_t messages_decoded = 0;

    while (in) {
        in.read(chunk.data(), chunk.size());
        const auto n = static_cast<std::size_t>(in.gcount());
        if (n == 0) break;
        buffer.insert(buffer.end(), chunk.begin(), chunk.begin() + n);

        auto entries = tensogram::scan(buffer.data(), buffer.size());
        std::size_t last_end = 0;
        for (const auto& e : entries) {
            const std::uint8_t* slice = buffer.data() + e.offset;
            auto msg = tensogram::decode(slice, e.length, {});
            auto obj = msg.object(0);
            auto meta = msg.get_metadata();
            ++messages_decoded;
            std::printf("  message %zu: param=%-5s  shape=[%zu,%zu]  %zu bytes\n",
                        messages_decoded,
                        meta.get_string("mars.param").c_str(),
                        obj.shape()[0], obj.shape()[1],
                        obj.data_size());
            last_end = e.offset + e.length;
        }
        if (last_end > 0) {
            // Drop consumed bytes; keep any partial-message tail.
            buffer.erase(buffer.begin(), buffer.begin() + last_end);
        }
    }

    std::printf("%s\n", std::string(60, '-').c_str());
    std::printf("Total: %zu messages decoded from stream\n", messages_decoded);
    std::printf("Peak buffer: single message (not whole file)\n");

    assert(messages_decoded == std::size(PARAMS));
    std::remove(tgm_path);
    return 0;
}
